### ISSUE TITLE
num_rows_subsample in DCRBaselineProtection / DCROverfittingProtection subsamples the real reference and removes the metrics' ability to detect memorization

### ISSUE THREAD

**Issue by kod5kod** [sdv-dev/SDMetrics#934]
### Environment
sdmetrics 0.31.0 (also reproduced on 0.30.0), Python 3.11, pandas 2.x.

### Description
Both DCR protection metrics accept `num_rows_subsample`. The docstring says it is "the number of
synthetic data rows to subsample from the synthetic data", and the documentation describes it as a
speed/precision trade-off. In `compute_breakdown` the real data is subsampled as well:

```python
if num_rows_subsample is not None:
    synthetic_sample = synthetic_data.sample(n=num_rows_subsample)
    random_sample = random_data.sample(n=num_rows_subsample)
    real_sample = real_data.sample(n=num_rows_subsample)   # <- reference side
```

Subsampling the *reference* of a nearest-neighbour query is not a precision loss; it is a bias in the
unsafe direction. A memorized synthetic row is only detectable while its source row is present in
the reference, so dropping reference rows inflates every DCR and the score moves toward "protected".
Averaging over `num_iterations` does not remove the bias, because every iteration has the same
expected inflation.

### Steps to reproduce
Frames with a known memorization rate (n = 5,000; 6 numerical + 3 categorical columns; `real` is a
sample from a Gaussian-copula model, `synth_leak_f` replaces a fraction `f` of an independent sample
with verbatim rows of `real`):

| f (fraction of synthetic rows copied from real) | 0 | 0.10 | 0.30 | 0.49 | 1.00 |
|---|---|---|---|---|---|
| DCRBaselineProtection, `num_rows_subsample=None` | 0.380 ± 0.015 | 0.362 ± 0.014 | 0.317 ± 0.014 | 0.183 ± 0.010 | 0.000 |
| DCRBaselineProtection, `num_rows_subsample=500`, 3 iterations | 0.455 ± 0.014 | 0.450 ± 0.009 | 0.440 ± 0.023 | 0.435 ± 0.021 | **0.419 ± 0.019** |
| DCROverfittingProtection, `num_rows_subsample=None` | 0.956 ± 0.043 | 0.901 ± 0.091 | 0.725 ± 0.106 | 0.530 ± 0.081 | 0.000 |
| DCROverfittingProtection, `num_rows_subsample=500`, 3 iterations | 0.941 ± 0.058 | 0.922 ± 0.055 | 0.927 ± 0.043 | 0.891 ± 0.039 | **0.886 ± 0.085** |

With subsampling at 500 rows, synthetic data that *is* the training set (f = 1.00) scores 0.419 /
0.886, against 0.455 / 0.941 for an independent sample with no memorization; for the overfitting metric
no leak fraction separates from the no-leak frame across 5 seeds.

### Minimal standalone reproduction (numpy + pandas + sdmetrics only)

Output on sdmetrics 0.31.0:
```
copied fraction 0.00:  no subsample = 0.218   num_rows_subsample=500 -> 0.269
copied fraction 0.50:  no subsample = 0.033   num_rows_subsample=500 -> 0.268
copied fraction 1.00:  no subsample = 0.000   num_rows_subsample=500 -> 0.257
```

```python
"""Standalone reproduction for the SDMetrics issue: needs only numpy, pandas, sdmetrics.

Real data is a correlated Gaussian sample; `synth` copies a fraction f of its rows verbatim and fills
the rest with an independent sample. Prints DCRBaselineProtection with and without subsampling.
"""

import numpy as np
import pandas as pd
from sdmetrics.single_table import DCRBaselineProtection

rng = np.random.default_rng(0)
n, d = 5000, 6
cov = 0.5 * np.ones((d, d)) + 0.5 * np.eye(d)
cols = [f"x{i}" for i in range(d)]
real = pd.DataFrame(rng.multivariate_normal(np.zeros(d), cov, n), columns=cols)
fresh = pd.DataFrame(rng.multivariate_normal(np.zeros(d), cov, n), columns=cols)
md = {"columns": {c: {"sdtype": "numerical"} for c in cols}}

for f in (0.0, 0.5, 1.0):
    k = int(f * n)
    synth = pd.concat([fresh.iloc[: n - k], real.iloc[rng.choice(n, k, replace=False)]]).sample(frac=1, random_state=0)
    full = DCRBaselineProtection.compute(real, synth, md, None)
    sub = DCRBaselineProtection.compute(real, synth, md, None, num_rows_subsample=500, num_iterations=3)
    print(f"copied fraction {f:.2f}:  no subsample = {full:.3f}   num_rows_subsample=500 -> {sub:.3f}")
```

### Expected behavior
Subsample only the synthetic (query) side, which costs variance but not bias, and keep the full
real data as the reference; or, if the reference must be capped for memory, document that the score
is then biased upward and by roughly how much. A separate observation, not a bug: both metrics
reduce by the median, so a memorized minority below 50% of rows moves the score very little at any
setting (0.368 → 0.351 at f = 0.10 here).

### Context
Found while auditing third-party privacy metrics with a ground-truth harness (FideliTab,
github.com/kod5kod/FideliTab); the same bias existed in FideliTab's own metrics before 0.5.0. Happy
to open a PR that splits the caps if that is welcome.
